### PR TITLE
Stm32 i2c timing option

### PR DESCRIPTION
**Pull request by LMESTM**
This change offers an additional option for configuring the exact timing of I2C V2 peripherals of STM32s.

The timing computation can either be left to the simple algorithm already in place in the driver, or can be computed through SW tools like STM32CubeMX and provided as an entry in a board's Device Tree file. 

This would look like:
```
 &i2c1 {
 	status = "okay";
	clock-frequency = <I2C_BITRATE_FAST>;
	timing = <0x10C01C20>;
 };
```

### CODE REVIEW

File: drivers/i2c/i2c_ll_stm32.c
Comment (gmarull) [inline drivers/i2c/i2c_ll_stm32.c:306]: Use `DT_PROP_OR` for embedded condition

File: drivers/i2c/i2c_ll_stm32_v2.c
Comment (gmarull) [inline drivers/i2c/i2c_ll_stm32_v2.c:610]: wrong indentation

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (gmarull) [inline dts/bindings/i2c/st,stm32-i2c-v2.yaml:20]: Maybe a better name would be `timingr`. Would also be nice to include some dts-bindings helpers to, e.g.

```
timingr = <(I2C_STM32_SCLL(x) | I2C_STM32_SCLH(y) | ...)>;
```

Comment (LMESTM) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml:20]: The way I see things,  users would mostly rely on STM32CubeMX (PC SW tool) which computes the timing value based on peripheral clocks and fall and rise time values. In STM32CubeMX the setting is described as timing so I tried to keep namings aligned.

File: drivers/i2c/i2c_ll_stm32.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.c]: thanks for the hint - this is easier to read now !

File: drivers/i2c/i2c_ll_stm32_v2.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32_v2.c:610]: fixed

**Review by pabigot — CHANGES_REQUESTED** [review]
I don't like the way this silently overrides any runtime configuration by passing flags like `I2C_SPEED_FAST` to the device that has this flags.  This seems likely to be confusing and to cause problems when different bus devices require different speeds.

A better approach might be to provide an array that specifies a map from a standard speed (100 kHz, 400 kHz, etc) to the desired value to use when programming that selected speed, assuming that the clock configuration is not likely to be runtime controllable.

I'm also a little concerned with the degree to which v1 and v2 STM I2C drivers diverge in functionality, but maybe that's unavoidable.

There is still interest in addressing #4039, which might be a better way to deal with this, though it requires some API redesign.

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (erwango) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml:20]: > Based on the use in the next commit I suggest ll-i2c-timing

Can you elaborate on the `ll` prefix? If this relates to the STM32Cube API in use, I'm not sure this a good choice, as this is a pure SW choice while timingr is a register of the V2 I2C IP (I2C_TIMINGR), which needs to be configured whatever the API used (HAL, LL, CMSIS, ...).
Also, IMHO, `-i2c-` feels a bit redundant as for a i2c binding parameter.

Comment (LMESTM) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml:20]: > I don't like the way this silently overrides any runtime configuration by passing flags like `I2C_SPEED_FAST` to the device that has this flags. This seems likely to be confusing and to cause problems when different bus devices require different speeds.

Right ... the timing option allows very precise and fine-tuned configuration for a complete and definitive HW setup, but it is not dynamic at all. The use case it targets is rather a final product than a development kit.

> A better approach might be to provide an array that specifies a map from a standard speed (100 kHz, 400 kHz, etc) to the desired value to use when programming that selected speed, assuming that the clock configuration is not likely to be runtime controllable.

That would handle a few more cases indeed, but still would not handle a runtime change of the clock tree. 
The default algorithm should rather be used for runtime configuration. 

> I'm also a little concerned with the degree to which v1 and v2 STM I2C drivers diverge in functionality, but maybe that's unavoidable.
v1 and v2 are quite different. This very fine timing capability (together with programming complexity) is V2 only. 

> There is still interest in addressing #4039, which might be a better way to deal with this, though it requires some API redesign.

This timing option was not primarily meant to allow new frequencies, but to allow a product designer to fine-tune the I2C values to precisely match his hardware with measured fall time / rise time . Even with #4039 introduction, this aspects would probably require introduction of a timing option. But then if #4039 is addressed  ... that would become impossible to provide a table of values as suggested earlier ...  

so do you think I should implement the support for 3 values (100/400/1000kHz) as suggested earlier ?

**Comment by LMESTM** [comment]
@pabigot thanks for your feedback. I have added a few answers to your points, and I think I would need some more guidance before deciding what to do next.

> I don't like the way this silently overrides any runtime configuration by passing flags like `I2C_SPEED_FAST` to the device that has this flags. This seems likely to be confusing and to cause problems when different bus devices require different speeds.

Right ... the timing option allows very precise and fine-tuned configuration for a complete and definitive HW setup, but it is not dynamic at all. The use case it targets is rather a final product than a development kit.

> A better approach might be to provide an array that specifies a map from a standard speed (100 kHz, 400 kHz, etc) to the desired value to use when programming that selected speed, assuming that the clock configuration is not likely to be runtime controllable.

That would handle a few more cases indeed, but still would not handle a runtime change of the clock tree. 
The default algorithm should rather be used for runtime configuration. 

> I'm also a little concerned with the degree to which v1 and v2 STM I2C drivers diverge in functionality, but maybe that's unavoidable.
v1 and v2 are quite different. This very fine timing capability (together with programming complexity) is V2 only. 

> There is still interest in addressing #4039, which might be a better way to deal with this, though it requires some API redesign.

This timing option was not primarily meant to allow new frequencies, but to allow a product designer to fine-tune the I2C values to precisely match his hardware with measured fall time / rise time . Even with #4039 introduction, this aspects would probably require introduction of a timing option. But then if #4039 is addressed  ... that would become impossible to provide a table of values as suggested earlier ...  

so do you think I should implement the support for 3 values (100/400/1000kHz) as suggested earlier ?

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (pabigot) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml:20]: > > Based on the use in the next commit I suggest ll-i2c-timing
> 
> Can you elaborate on the `ll` prefix?

The value is passed directly to `LL_I2C_SetTiming()` in the driver.  That underneath that assigns `TIMINGR` in the I2C peripheral register set is true, but nobody will see that unless they look at the HAL implementation.

That there are other APIs than the one used is a good point.  `i2c-` because, again, that's part of the macros and constants used to construct and use the value, but I can see a case for assuming the context is the I2C peripheral structure and eliminating it.  

This isn't a blocking item, so I'll defer to @erwango.  "timingr" still seems a little opaque, but that could be addressed in the property description.

**Comment by pabigot** [comment]
> Right ... the timing option allows very precise and fine-tuned configuration for a complete and definitive HW setup, but it is not dynamic at all. The use case it targets is rather a final product than a development kit.

That's a worthwhile use case, but the limitations should be made extremely clear in the description so new folks don't shoot themselves.

> That would handle a few more cases indeed, but still would not handle a runtime change of the clock tree.

How likely is such a change?  And how difficult is it to detect?

> if #4039 is addressed ... that would become impossible to provide a table of values as suggested earlier ...

I would envision some sort of back door, e.g. not restricting the speed cell of the table to match a standard rate.

> so do you think I should implement the support for 3 values (100/400/1000kHz) as suggested earlier ?

My preference is a table like:
```
timingr = <100 0x4std>,  <400 0x4fast>, <1000 0x4fast+>;
```
collecting a sequence of 2-cell assignments mapping bus speed in kHz to the TIMINGR register value.  There's no requirement that the table be complete, nor that the speeds be standard speeds.  Provide only the values required for the application.  If a speed requested at runtime is not present in the table then either fall back to runtime calculation or fail the configuration, depending on how robust you want your system to be.

As an extra, if it's possible to specify the clock configuration assumed in the precalculated values, and detect at runtime that it's not right, fallback to runtime or fail for that too makes sense.

I believe the table approach has more potential than the current solution for being extensible to other SoCs.  Would that be acceptable?

**Comment by LMESTM** [comment]
> I believe the table approach has more potential than the current solution for being extensible to other SoCs. Would that be acceptable?

Yes I hope so. You proposal with fallback solutions seems the more appropriate. To cover most cases, I guess we would need tables with 3 values per entry: clock source frequency, I2C requested frequency,  and timing value. 
timingr = <64000000 100 0x4std>,  <64000000 400 0x4fast>, <64000000 1000 0x4fast+>;
Then we can check the 2 first ones and decide to apply timing or fallback. 

I have to check how to get table with dynamic lengths from device tree and how to allocate memory for storing it.

**Comment by LMESTM** [comment]
@erwango @gmarull @pabigot  I followed latest recommendations on providing a list of possible settings with their matching clock configuration. I came-up with a new proposal that seem to be flexible enough.  Let me know your feedback.

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (pabigot) [inline dts/bindings/i2c/st,stm32-i2c-v2.yaml:20]: The `I2C_BITRATE_` devicetree macros specify speed in Hz, not kHz.

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (pabigot) [inline dts/bindings/i2c/st,stm32-i2c-v2.yaml:17]: Isn't this still `TIMINGR`?

File: drivers/i2c/i2c_ll_stm32.h
Comment (pabigot) [inline drivers/i2c/i2c_ll_stm32.h:19]: Please document units at each of these (I think `Hz` for the first two, and `register value` for the third).

File: drivers/i2c/i2c_ll_stm32_v2.c
Comment (pabigot) [inline drivers/i2c/i2c_ll_stm32_v2.c:609]: Indentation problem somewhere.

**Review by pabigot — COMMENTED** [review]
I think this is going to work; a few problems noted below.

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (LMESTM) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml]: updated in last PR version

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (LMESTM) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml]: right - fixed

File: drivers/i2c/i2c_ll_stm32_v2.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32_v2.c]: hopefully fixed - checkpatch should confirm

File: drivers/i2c/i2c_ll_stm32.h
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.h:20]: done

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (erwango) [inline dts/bindings/i2c/st,stm32-i2c-v2.yaml:21]: This details should be made part of the property description.

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (erwango) [inline dts/bindings/i2c/st,stm32-i2c-v2.yaml:20]: Let's re-use the official i2c speed binding name here:
```suggestion
    # The resulting table entries should look like <periph_clock clock-frequency timing>
```

File: drivers/i2c/i2c_ll_stm32.h
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.h:19]:
```suggestion
	/* i2c peripheral clock in Hz */
```

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:295]: We're trying to get limit the use of Kconfig flags and replace by compatibles whenever it is possible,
so here it would be:
```suggestion
COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(i2c_ll_stm32_v2),					\
```
However, I'm not sure this will work well, specially in the header file.
So if working, this would be nice, but this is not a blocking comment.

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:308]: If possible, cf previous comment:
```suggestion
	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(i2c_ll_stm32_v2),				\
```

File: drivers/i2c/i2c_ll_stm32_v2.c
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32_v2.c:613]: Non blocking: 'preset' would be clearer than 'fixed'

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (LMESTM) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml:21]: I haven't found a description that is longer than a small line.
Should I put everything inside the description ?  is it allowed to have multiple lines ?

Comment (erwango) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml:21]: Yes, multiple lines are allowed and it can be formatted nicely:
https://github.com/zephyrproject-rtos/zephyr/pull/25996/files#diff-ff68241e84846fc765b7a5d0e2c6bcc8

File: drivers/i2c/i2c_ll_stm32.h
Comment (gmarull) [inline drivers/i2c/i2c_ll_stm32.h:37]: type should be `size_t`, I'd suggest shorter name, e.g. `n_timings`

File: drivers/i2c/i2c_ll_stm32.h
Comment (gmarull) [inline drivers/i2c/i2c_ll_stm32.h:36]: `_list` prefix is redundant as `timings` is plural, implying > 1.

File: drivers/i2c/i2c_ll_stm32.h
Comment (gmarull) [inline drivers/i2c/i2c_ll_stm32.h:24]: maybe just name `timingr`?

File: drivers/i2c/i2c_ll_stm32.c
Comment (gmarull) [inline drivers/i2c/i2c_ll_stm32.c:309]: why is cast needed here?

File: drivers/i2c/i2c_ll_stm32.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.c:295]: Just tested this, and it doesn't work : it does compile, but the condition seems to be evaluated to false

File: drivers/i2c/i2c_ll_stm32.h
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.h:19]: thx will fix

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (LMESTM) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml:20]: will do

File: drivers/i2c/i2c_ll_stm32_v2.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32_v2.c:613]: will do

File: drivers/i2c/i2c_ll_stm32.h
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.h:37]: done

File: drivers/i2c/i2c_ll_stm32.h
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.h:36]: and that will make lines even shorter - done

File: drivers/i2c/i2c_ll_stm32.h
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.h:24]: I know that this is the register name but I just felt that timingr is not meaningful in the code and structure  .... though if everyone agrees, I can do so

File: drivers/i2c/i2c_ll_stm32.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.c:309]: i2c_timings_##name is a table of uint32_t  retrieved from the device tree array.
Even though the device tree array is of form <a b>, <c d> it results in a table of {a, b, c, d}, not {{a, b}, {c, d}} 
therefore I cannot make i2c_timings_##name a table of const struct i2c_config_timing - I tried and I would get a warning. so I decided to get this uint32_t table and cast it to the our structured format.

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (LMESTM) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml]: done

**Comment by LMESTM** [comment]
updated PR sent with fixes I could do. I left few comments open when I did not implement the change - let me know if more is needed.

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c:295]: ok, we should find a way to do this better then.
Let's keep it as is, txs for the attempt

File: drivers/i2c/i2c_ll_stm32.h
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.h:24]: I don't have a string opinion here, @pabigot maybe ?

**Review by erwango — APPROVED** [review]

File: drivers/i2c/i2c_ll_stm32.c
Comment (pabigot) [inline drivers/i2c/i2c_ll_stm32.c:295]: You can make this much more readable (and consistent with other code) as:
```
#if DT_HAS_COMPAT_STATUS_OKAY(i2c_ll_stm32_v2)
#define DEFINE_TIMINGS(name)						\
	static const uint32_t i2c_timings_##name[] =			\
		DT_PROP_OR(DT_NODELABEL(name), timings, {});
#define USE_TIMINGS(name)						\
	.timings = (const struct i2c_config_timing *) i2c_timings_##name, \
	.n_timings = ARRAY_SIZE(i2c_timings_##name),
#else /* V2 */
#define DEFINE_TIMINGS(name)
#define USE_TIMINGS(name)
#endif /* V2 */

#define STM32_I2C_INIT(name)						\
STM32_I2C_IRQ_HANDLER_DECL(name);					\
									\
DEFINE_TIMINGS(name)							\
									\
static const struct i2c_stm32_config i2c_stm32_cfg_##name = {		\
	.i2c = (I2C_TypeDef *)DT_REG_ADDR(DT_NODELABEL(name)),		\
	.pclken = {							\
		.enr = DT_CLOCKS_CELL(DT_NODELABEL(name), bits),	\
		.bus = DT_CLOCKS_CELL(DT_NODELABEL(name), bus),		\
	},								\
	STM32_I2C_IRQ_HANDLER_FUNCTION(name)				\
	.bitrate = DT_PROP(DT_NODELABEL(name), clock_frequency),	\
	USE_TIMINGS(name)						\
};									\
```

However, whatever preprocessor expression is using to select this must match what's done in `i2c_ll_stm32.h`.

File: drivers/i2c/i2c_ll_stm32.h
Comment (pabigot) [reply drivers/i2c/i2c_ll_stm32.h:24]: I don't have standing to make the decision, just raise the concern.  By my process if the SOC maintainer @erwango defers, then it defers to the implementer @LMESTM.  I'm fine with `timing_setting`.

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (pabigot) [inline dts/bindings/i2c/st,stm32-i2c-v2.yaml:21]: I'd really like to see this more readable in plain text in the yaml, which is where I most often look at it.  That includes line length and paragraph clarity.  This should work:
```
        description: |
            An optional table of pre-computed i2c timing values with the
            matching clock configuration.

            Precise timings values for a given Hardware can be pre-computed
            with a tool like STM32CubeMX or directly from I2C_TIMINGR register
            description.

            Because timing value is valid for a given I2C peripheral clock
            frequency and target I2C bus clock, each timing value must be
            provided with the matching configuration.

            The resulting table entries should look like: <periph_clock
            clock-frequency timing>

            For example timings could be defined as

                timings = <64000000 I2C_BITRATE_STANDARD 0x10707DBC>,
                          <64000000 I2C_BITRATE_FAST 0x00603D56>,
                          <56000000 I2C_BITRATE_STANDARD 0x10606DA4>,
                          <56000000 I2C_BITRATE_FAST 0x00501D63>;
```
and it also gives the hint that the devicetree content should be multiline (rather than going off past the horizon).

(You can't put a `:` at the end of the `For example`, but that's yaml for you....)

**Review by pabigot — CHANGES_REQUESTED** [review]
Two changes please.

File: drivers/i2c/i2c_ll_stm32.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.c]: @gmarull does that answer your question  ? would you have a alternative proposal ?

File: drivers/i2c/i2c_ll_stm32.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.c:295]: Ok for the proposals, thanks for providing the code.
Just one point, from the tests I run:
#if DT_HAS_COMPAT_STATUS_OKAY(i2c_ll_stm32_v2)  => does not work (condition is not met)
while 
#ifdef CONFIG_I2C_STM32_V2   => works OK
@erwango and @pabigot both think that it should work, am I missing something ? 
I'm on master form Aug 19th, is this something that has been introduced in the past few days ?

File: dts/bindings/i2c/st,stm32-i2c-v2.yaml
Comment (LMESTM) [reply dts/bindings/i2c/st,stm32-i2c-v2.yaml:21]: thanks for formatting - will use  it

File: drivers/i2c/i2c_ll_stm32.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.c]: For now, I have pushed the format changes except I use #ifdef CONFIG_I2C_STM32_V2

File: drivers/i2c/i2c_ll_stm32.c
Comment (gmarull) [reply drivers/i2c/i2c_ll_stm32.c]: Got it, thanks for the clarification

File: drivers/i2c/i2c_ll_stm32.c
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c]: @LMESTM during your trial, are you sure you have a matching dts node enabled ?

Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.c]: > @LMESTM during your trial, are you sure you have a matching dts node enabled ?

Well the timings handling is only done in I2C v2 driver. 
As said, timing is applied / not applied depending on the change below only: 
#if DT_HAS_COMPAT_STATUS_OKAY(i2c_ll_stm32_v2)
#ifdef CONFIG_I2C_STM32_V2 => works OK

Am I expect the driver type to be defined by default, do I need to add something in DT to get i2c_ll_stm32_v2 defined ?

Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.c]: ok I found out the proper form: 
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_i2c_v2)
and I pushed a new update accordingly

File: drivers/i2c/i2c_ll_stm32.c
Comment (gmarull) [inline drivers/i2c/i2c_ll_stm32.c:299]: #else case is for `V1`?

**Review by gmarull — APPROVED** [review]
LGTM

File: drivers/i2c/i2c_ll_stm32.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.c:299]: yes - this timing option is not applicable for v1

File: drivers/i2c/i2c_ll_stm32.h
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.h:14]:
```suggestion
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_i2c_v2)
```

File: drivers/i2c/i2c_ll_stm32.h
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.h:35]:
```suggestion
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_i2c_v2)
```

**Review by erwango — CHANGES_REQUESTED** [review]
2 last changes and we're good

File: drivers/i2c/i2c_ll_stm32.h
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.h]: I missed those 2 ones indeed, thanks @erwango

File: drivers/i2c/i2c_ll_stm32.h
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32.h]: change done

File: drivers/i2c/i2c_ll_stm32_v2.c
Comment (pabigot) [inline drivers/i2c/i2c_ll_stm32_v2.c:606]: One last nit: please don't put parentheses around `cfg->n_timings`.

**Review by pabigot — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]

File: drivers/i2c/i2c_ll_stm32_v2.c
Comment (LMESTM) [reply drivers/i2c/i2c_ll_stm32_v2.c]: done as well

**Review by pabigot — APPROVED** [review]

**Pull request merged**
